src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,345 @@
+
+        if hasattr(model.config, "use_cache"):
+            assert not model.config.use_cache, "`use_cache` must be False"
+
+        use_trainable_adapters = peft_config.use_trainable_adapters
+        adapters_items = iter(peft_config.adapters.items())
+
+        # Because we call load_adapter, which requires base_model to be defined
+        model_peft.base_model = self
+        # For load_adapter to think we are a LoraModel
+        model_peft.peft_type = PeftType.LORA
+
+        for adapter_name, model_id in adapters_items:
+            model_peft.load_adapter(model_id, adapter_name, is_trainable=use_trainable_adapters)
+
+        self.set_adapter(list(peft_config.adapters.keys()))
+        model_peft.peft_type = PeftType.XLORA
+
+        def hook(module, *args, **kwargs) -> None:
+            args_real = args[0]
+            kwargs_real: dict = args[1]
+            kwargs_real.update(kwargs)
+
+            xlora_classifier: xLoRAClassifier = self.internal_xlora_classifier  # type: ignore
+
+            if "_xlora_classifier_inhibitor_flag" in kwargs_real:
+                payload: InhibitorFlagPayload = kwargs_real["_xlora_classifier_inhibitor_flag"]
+
+                del kwargs_real["_xlora_classifier_inhibitor_flag"]
+
+                self.internal_xlora_scalings = torch.full(  # type: ignore
+                    (payload.batch_size, payload.seq_len, xlora_classifier.n_layers, xlora_classifier.n_classes),
+                    payload.override_scaling_pass_value,
+                )
+
+                return
+
+            xlora_scalings = xlora_classifier.forward(
+                *args_real,
+                **kwargs_real,
+            )
+            # Set the scalings
+            self.internal_xlora_scalings = xlora_scalings
+
+        model.register_forward_pre_hook(hook, with_kwargs=True, prepend=True)
+
+        self.eval()
+        if not use_trainable_adapters:
+            total_frozen = 0
+            for name, param in self.named_parameters():
+                if "lora_" in name:
+                    param.requires_grad = False
+                    total_frozen += 1
+
+        assert isinstance(self, LoraModel)
+
+        total_swapped = convert_layers_to_xlora(
+            model_peft,
+            peft_config,
+        )
+
+        n_classes = len(peft_config.adapters)
+        xlora_classifier = xLoRAClassifier(model_peft, peft_config, n_classes, total_swapped)
+
+        # Setup the internal state
+        base_model_wrapper = BaseTunerWrapper(self, xlora_classifier)
+        self.forward = base_model_wrapper.forward  # type: ignore[method-assign]
+
+        peft_model_wrapper = PeftModelWrapper(
+            model_peft,
+            model_peft.save_pretrained,
+            peft_config,
+            model_peft.get_nb_trainable_parameters,
+            model_peft.generate,
+        )
+        model_peft.save_pretrained = peft_model_wrapper.save_pretrained  # type: ignore
```

Comment:
Not a fan of assigning methods like this, would it be possible to define them normally?

Reply:
I'm not sure if there is another way, I think that method assignment through a reference to `PeftModel` is the only way to modify the methods? We could create a `XLoraPeftModel`, but that would defeat the purpose of having a tuner class.

Reply:
I don't think I fully understand this yet. Above, we call:

`self.base_model.__xlora_post_init__(model, peft_config, adapter_name)`

but here it looks like this method requires 4 arguments. What is the `model_peft`, is this the `PeftModel`?

So what I suspect is happening here is that you want to modify the behavior of `generate` and `save_pretrained` of the `PeftModel` without actually modifying `PeftModel`, is this the goal?

When it comes to `generate`, the `PeftModel` calls `generate` on the `base_model` anyway, could we not add the modification at that point?

When it comes to `save_pretrained`, I think we could check if we can somehow make the two approaches work together, I'd need to understand what the custom `save_pretrained` method does differently.

One easy way that we could allow custom `save_pretrained` methods without a lot of changes in `PeftModel` would be something like this:

```python
class PeftModel:
    def save_pretrained(...):
        if hasattr(self.base_model, "_custom_save_pretrained"):
            return self.base_model._custom_save_pretrained(...)
        # same code as right now
```

This way, we only added 2 extra lines in `PeftModel` but would allow the underlying model to implement their own `save_pretrained` method.

Reply:
> but here it looks like this method requires 4 arguments. What is the model_peft, is this the PeftModel?

This was a mistake, it should be 3 arguments. The model_peft should be typed as a PeftModel (and the `xlora` lib has it as such), but due to circular imports I made it `nn.Module` and left a comment.

> So what I suspect is happening here is that you want to modify the behavior of generate and save_pretrained of the PeftModel without actually modifying PeftModel, is this the goal?

Yes, this approach is because we wrote the code outside of `peft`, and could not change the necessary methods. 

> When it comes to generate, the PeftModel calls generate on the base_model anyway, could we not add the modification at that point?

That sounds great! I have implemented it. I think we could also do a similar thing with `save_pretrained`, as you outlined.

> When it comes to save_pretrained, I think we could check if we can somehow make the two approaches work together, I'd need to understand what the custom save_pretrained method does differently.

With `save_pretrained` what we do differently is that we save the adapters in a specific directory (`save_directory+"/adapters"` or the like), in addition to the config file and the classifier. To avoid code duplication, we could change the saving spot (this would be simple) and implement a method such as what you outlined above to house the config and classifier saving mechanics.

I implemented a hook as you outlined in [this](https://github.com/huggingface/peft/pull/1491/commits/1b5175e63038e13227689791d9a0695456e82264) commit. The `PeftModelWrapper` class does not exist anymore as all the methods have been moved out.